contrib/i2pd.service

```diff
@@ -2,31 +2,60 @@
 Description=I2P Router written in C++
 Documentation=man:i2pd(1) https://i2pd.readthedocs.io/en/latest/
 After=network.target
+ConditionFileIsExecutable=/usr/sbin/i2pd
 
 [Service]
 User=i2pd
 Group=i2pd
+PermissionsStartOnly=yes
+
 RuntimeDirectory=i2pd
 RuntimeDirectoryMode=0700
 LogsDirectory=i2pd
 LogsDirectoryMode=0700
-Type=forking
-ExecStart=/usr/sbin/i2pd --conf=/etc/i2pd/i2pd.conf --tunconf=/etc/i2pd/tunnels.conf --tunnelsdir=/etc/i2pd/tunnels.conf.d --pidfile=/run/i2pd/i2pd.pid --logfile=/var/log/i2pd/i2pd.log --daemon --service
+Type=exec
+Environment="DAEMON_OPTS=--conf=/etc/i2pd/i2pd.conf --tunconf=/etc/i2pd/tunnels.conf --tunnelsdir=/etc/i2pd/tunnels.conf.d --pidfile=/run/i2pd/i2pd.pid --logfile=/var/log/i2pd/i2pd.log --service"
+EnvironmentFile=-/etc/default/i2pd
+ExecStart=/usr/sbin/i2pd $DAEMON_OPTS
 ExecReload=/bin/sh -c "kill -HUP $MAINPID"
 PIDFile=/run/i2pd/i2pd.pid
-### Uncomment, if auto restart needed
-#Restart=on-failure
+Restart=on-failure
 
 KillSignal=SIGQUIT
 # If you have the patience waiting 10 min on restarting/stopping it, uncomment this.
 # i2pd stops accepting new tunnels and waits ~10 min while old ones do not die.
 #KillSignal=SIGINT
 #TimeoutStopSec=10m
 
-# If you have problems with hanging i2pd, you can try increase this
-LimitNOFILE=4096
+LimitNOFILE=65536
 # To enable write of coredump uncomment this
 #LimitCORE=infinity
 
+# Hardening
+NoNewPrivileges=yes
+PrivateTmp=yes
+PrivateUsers=true
+PrivateDevices=yes
+ProtectHome=yes
+ProtectSystem=full
+ProtectClock=true
+ProtectControlGroups=yes
+ProtectHostname=true
+ProtectKernelLogs=true
+ProtectKernelModules=yes
+ProtectKernelTunables=yes
+ProtectSystem=strict
+RestrictAddressFamilies=AF_INET AF_INET6
+RestrictNamespaces=cgroup ipc mnt pid user uts
+RestrictRealtime=true
+RestrictSUIDSGID=true
+RemoveIPC=true
+ReadOnlyDirectories=/
+ReadWriteDirectories=-/proc
```

Comment:
why read and write to `/proc`? shouldn't this be at most read only?

Reply:
It **_seems_** like it works fine with `/proc` as read only. However I don't see why not allow it full access to `/proc`? 

After all it's going to be able to write in `/proc`  to it's own i2p user only, so read-only and read-write wouldn't make any meaningful difference security-wise.  The plus is that we minimize chances of something going wrong.

Reply:
full access to `/proc` is similar in nature to full access to `/sys` both can be abused to possibly exploit a theoretical privesc bug. also general practice is to allow only what is needed and nothing more.

Reply:
`/proc` line got removed.

Reply:
Is there an option to disable daemon? Something like `--no-daemon`?

Older default debian configurations had `daemon=true` but thankfully this is no longer default.

Reply:
This doesn't seem true. My bad